File: src/test/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/integrasjoner/enhetsregisteret/EnhetsregisteretClientTest.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.enhetsregisteret;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import lombok.SneakyThrows;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Orgnr;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.OverordnetEnhet;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Underenhet;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.HttpStatus;
import org.springframework.web.client.HttpServerErrorException;
import org.springframework.web.client.RestTemplate;

import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.TestData.etOrgnr;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class EnhetsregisteretClientTest {

    private EnhetsregisteretClient enhetsregisteretClient;
    private final static ObjectMapper objectMapper = new ObjectMapper();

    @Mock
    private RestTemplate restTemplate;

    @BeforeEach
    public void setup() {
        enhetsregisteretClient = new EnhetsregisteretClient(restTemplate, "");
    }

    @Test
    public void hentInformasjonOmEnhet__skal_hente_riktige_felter() {
        mockRespons(gyldigEnhetRespons("999263550"));
        OverordnetEnhet overordnetEnhet = enhetsregisteretClient.hentInformasjonOmEnhet(new Orgnr("999263550"));

        assertThat(overordnetEnhet.getOrgnr().getVerdi()).isEqualTo("999263550");
        assertThat(overordnetEnhet.getNavn()).isEqualTo("NAV ARBEID OG YTELSER");
        assertThat(overordnetEnhet.getNæringskode().getKode()).isEqualTo("84300");
        assertThat(overordnetEnhet.getNæringskode().getBeskrivelse()).isEqualTo("Trygdeordninger underlagt offentlig forvaltning");
        assertThat(overordnetEnhet.getInstitusjonellSektorkode().getKode()).isEqualTo("6100");
        assertThat(overordnetEnhet.getInstitusjonellSektorkode().getBeskrivelse()).isEqualTo("Statsforvaltningen");
        assertThat(overordnetEnhet.getAntallAnsatte()).isEqualTo(40);
    }

    @Test
    public void hentInformasjonOmEnhet__skal_feile_hvis_server_returnerer_5xx_server_returnerer_5xx() {
        when(
                restTemplate.getForObject(
                        anyString(),
                        any()
                )
        ).thenThrow(
                new HttpServerErrorException(HttpStatus.BAD_GATEWAY)
        );

        assertThrows(EnhetsregisteretException.class, () -> enhetsregisteretClient.hentInformasjonOmEnhet(etOrgnr()));
    }

    @Test
    public void hentInformasjonOmEnhet__skal_feile_hvis_et_felt_mangler() {
        ObjectNode responsMedManglendeFelt = gyldigEnhetRespons("999263550");
        responsMedManglendeFelt.remove("institusjonellSektorkode");
        mockRespons(responsMedManglendeFelt);

        assertThrows(EnhetsregisteretMappingException.class, () -> enhetsregisteretClient.hentInformasjonOmEnhet(etOrgnr()));
    }

    @Test
    public void hentInformasjonOmEnhet__skal_feile_hvis_returnert_orgnr_ikke_matcher_med_medsendt_orgnr() {
        ObjectNode responsMedFeilOrgnr = gyldigEnhetRespons("999263550");
        mockRespons(responsMedFeilOrgnr);
        assertThrows(
                IllegalStateException.class,
                () -> enhetsregisteretClient.hentInformasjonOmEnhet(new Orgnr("777777777"))
        );
    }

    @Test
    public void hentInformasjonOmUnderenhet__skal_hente_riktige_felter() {
        mockRespons(gyldigUnderenhetRespons("971800534"));
        Underenhet underenhet = enhetsregisteretClient.hentUnderenhet(new Orgnr("971800534"));

        assertThat(underenhet.getOrgnr().getVerdi()).isEqualTo("971800534");
        assertThat(underenhet.getOverordnetEnhetOrgnr().getVerdi()).isEqualTo("999263550");
        assertThat(underenhet.getNavn()).isEqualTo("NAV ARBEID OG YTELSER AVD OSLO");
        assertThat(underenhet.getNæringskode().getKode()).isEqualTo("84300");
        assertThat(underenhet.getNæringskode().getBeskrivelse()).isEqualTo("Trygdeordninger underlagt offentlig forvaltning");
        assertThat(underenhet.getAntallAnsatte()).isEqualTo(40);
    }

    @Test
    public void hentInformasjonOmUnderenhet__skal_feile_hvis_server_returnerer_5xx() {
        when(
                restTemplate.getForObject(
                        anyString(),
                        any()
                )
        ).thenThrow(
                new HttpServerErrorException(HttpStatus.BAD_GATEWAY)
        );


        assertThrows(
                EnhetsregisteretIkkeTilgjengeligException.class,
                () -> enhetsregisteretClient.hentUnderenhet(etOrgnr())
        );
    }
    @Test
    public void hentInformasjonOmUnderenhet__skal_feile_hvis_et_felt_mangler() {
        ObjectNode responsMedManglendeFelt = gyldigUnderenhetRespons("822565212");
        responsMedManglendeFelt.remove("navn");
        mockRespons(responsMedManglendeFelt);

        assertThrows(
                EnhetsregisteretMappingException.class,
                () -> enhetsregisteretClient.hentUnderenhet(etOrgnr())
        );
    }

    @Test
    public void hentInformasjonOmUnderenhet__skal_feile_hvis_returnert_orgnr_ikke_matcher_med_medsendt_orgnr() {
        ObjectNode responsMedFeilOrgnr = gyldigUnderenhetRespons("822565212");
        mockRespons(responsMedFeilOrgnr);

        assertThrows(
                IllegalStateException.class,
                () -> enhetsregisteretClient.hentUnderenhet(new Orgnr("777777777"))
        );
    }

    @SneakyThrows
    private void mockRespons(JsonNode node) {
        when(restTemplate.getForObject(anyString(), any())).thenReturn(objectMapper.writeValueAsString(node));
    }

    @SneakyThrows
    private ObjectNode gyldigUnderenhetRespons(String orgnr) {
        String str = "{\n" +
                "  \"organisasjonsnummer\": \"" + orgnr + "\",\n" +
                "  \"navn\": \"NAV ARBEID OG YTELSER AVD OSLO\",\n" +
                "  \"naeringskode1\": {\n" +
                "    \"beskrivelse\": \"Trygdeordninger underlagt offentlig forvaltning\",\n" +
                "    \"kode\": \"84.300\"\n" +
                "  },\n" +
                "  \"antallAnsatte\": 40,\n" +
                "  \"overordnetEnhet\": \"999263550\"\n" +
                "}";
        return (ObjectNode) objectMapper.readTree(str);
    }

    @SneakyThrows
    private ObjectNode gyldigEnhetRespons(String orgnr) {
        String str = "{\n" +
                "  \"organisasjonsnummer\": \"" + orgnr + "\",\n" +
                "  \"navn\": \"NAV ARBEID OG YTELSER\",\n" +
                "  \"naeringskode1\": {\n" +
                "    \"beskrivelse\": \"Trygdeordninger underlagt offentlig forvaltning\",\n" +
                "    \"kode\": \"84.300\"\n" +
                "  },\n" +
                "  \"institusjonellSektorkode\": {\n" +
                "    \"kode\": \"6100\",\n" +
                "    \"beskrivelse\": \"Statsforvaltningen\"\n" +
                "  },\n" +
                "  \"antallAnsatte\": 40\n" +
                "}";
        return (ObjectNode) objectMapper.readTree(str);
    }
}

File: src/test/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/integrasjoner/enhetsregisteret/EnhetsregisteretClientTest.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.enhetsregisteret

import arrow.core.left
import com.github.tomakehurst.wiremock.client.WireMock.*
import com.github.tomakehurst.wiremock.junit5.WireMockRuntimeInfo
import com.github.tomakehurst.wiremock.junit5.WireMockTest
import common.StaticAppenderExtension
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.*
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.fail
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.web.reactive.function.client.WebClient

@WireMockTest
@ExtendWith(StaticAppenderExtension::class)
class EnhetsregisteretClientTest {
    private val url get() = "http://localhost:${port}/enhetsregisteret"
    private val webClient = WebClient.create()
    private val enhetsregisteretClient: EnhetsregisteretClient by lazy {
        EnhetsregisteretClient(
            webClient, url
        )
    }

    @Test
    fun `hent informasjon om enhet skal hente riktig felter`() {
        stubFor(
            get("/enhetsregisteret/enheter/999263550").willReturn(
                okJson(
                    """
                    {
                      "organisasjonsnummer": "999263550",
                      "navn": "NAV ARBEID OG YTELSER",
                      "naeringskode1": {
                        "kode": "84300",
                        "beskrivelse": "Trygdeordninger underlagt offentlig forvaltning"
                      },
                      "institusjonellSektorkode": {
                        "kode": "6100",
                        "beskrivelse": "Statsforvaltningen"
                      },
                      "antallAnsatte": 40
                    }
                    """.trimIndent()
                )
            )
        )
        val overordnetEnhet = enhetsregisteretClient.hentEnhet(Orgnr("999263550"))

        assertThat(overordnetEnhet.getOrNull()).isEqualTo(
            OverordnetEnhet(
                orgnr = Orgnr("999263550"),
                navn = "NAV ARBEID OG YTELSER",
                næringskode = Næringskode5Siffer("84300", "Trygdeordninger underlagt offentlig forvaltning"),
                institusjonellSektorkode = InstitusjonellSektorkode("6100", "Statsforvaltningen"),
                antallAnsatte = 40
            )
        )
    }

    @Test
    fun `hent enhet skal feile hvis server returnerer 5xx`() {
        stubFor(
            get("/enhetsregisteret/enheter/999263550").willReturn(
                serverError()
            )
        )
        enhetsregisteretClient.hentEnhet(Orgnr("999263550")).fold(
            {
                assertThat(it).isEqualTo(EnhetsregisteretClient.HentEnhetFeil.FeilVedKallTilEnhetsregisteret)
            },
            {
                fail("Skulle returnert left")
            }
        )
    }

    @Test
    fun `hent enhet skal feile hvis et felt mangler i responsen`() {
        stubFor(
            get("/enhetsregisteret/enheter/999263550").willReturn(
                okJson(
                    """
                    {
                      "organisasjonsnummer": "999263550",
                      "navn": "NAV ARBEID OG YTELSER",
                      "naeringskode1": {
                        "kode": "84300",
                        "beskrivelse": "Trygdeordninger underlagt offentlig forvaltning"
                      },
                      "antallAnsatte": 40
                    }
                    """.trimIndent()
                )
            )
        )

        val result = enhetsregisteretClient.hentEnhet(Orgnr("999263550"))
        assertThat(result).isEqualTo(EnhetsregisteretClient.HentEnhetFeil.FeilVedDekodingAvJson.left())
    }

    @Test
    fun `hent enhet skal feile hvis returnert orgnr ikke matcher med medsendt orgnr`() {
        stubFor(
            get("/enhetsregisteret/enheter/777777777").willReturn(
                okJson(
                    """
                    {
                    "organisasjonsnummer": "999263550",
                    "navn": "NAV ARBEID OG YTELSER",
                    "naeringskode1": {
                        "kode": "84300",
                        "beskrivelse": "Trygdeordninger underlagt offentlig forvaltning"
                    },
                    "institusjonellSektorkode": {
                        "kode": "6100",
                        "beskrivelse": "Statsforvaltningen"
                    },
                    "antallAnsatte": 40
                    }
                    """.trimIndent()
                )
            )
        )
        enhetsregisteretClient.hentEnhet(Orgnr("777777777")).fold(
            {
                assertThat(it).isEqualTo(EnhetsregisteretClient.HentEnhetFeil.OrgnrMatcherIkke)
            },
            {
                fail("Skulle feilet")
            }
        )
    }

    @Test
    fun `hent underenheter skal hente riktige felter`() {
        stubFor(
            get("/enhetsregisteret/underenheter/971800534").willReturn(
                okJson(
                    """
                    {
                      "organisasjonsnummer": "971800534",
                      "navn": "NAV ARBEID OG YTELSER AVD OSLO",
                      "overordnetEnhet": "999263550",
                      "naeringskode1": {
                        "kode": "84300",
                        "beskrivelse": "Trygdeordninger underlagt offentlig forvaltning"
                      },
                      "antallAnsatte": 40
                    }
                    """.trimIndent()
                )
            )
        )

        val underenhet = enhetsregisteretClient.hentUnderenhet(Orgnr("971800534"))
        assertThat(underenhet.getOrNull()).isEqualTo(
            Underenhet.Næringsdrivende(
                orgnr = Orgnr("971800534"),
                navn = "NAV ARBEID OG YTELSER AVD OSLO",
                overordnetEnhetOrgnr = Orgnr("999263550"),
                næringskode = Næringskode5Siffer("84300", "Trygdeordninger underlagt offentlig forvaltning"),
                antallAnsatte = 40
            )
        )
    }

    @Test
    fun `hent underenhet skal feile hvis enhetsregisteret returnerer 5xx`() {
        stubFor(
            get("/enhetsregisteret/underenheter/971800534").willReturn(
                serverError()
            )
        )

        val result = enhetsregisteretClient.hentUnderenhet(Orgnr("971800534"))
        assertThat(result).isEqualTo(EnhetsregisteretClient.HentUnderenhetFeil.EnhetsregisteretSvarerIkke.left())
    }

    @Test
    fun `hent underenhet skal feile hvis ett felt mangler`() {
        stubFor(
            get("/enhetsregisteret/underenheter/971800534").willReturn(
                okJson(
                    """
                    {
                      "navn": "NAV ARBEID OG YTELSER AVD OSLO",
                      "overordnetEnhet": "999263550",
                      "naeringskode1": {
                        "kode": "84300",
                        "beskrivelse": "Trygdeordninger underlagt offentlig forvaltning"
                      }
                    }
                    """.trimIndent()
                )
            )
        )

        val result = enhetsregisteretClient.hentUnderenhet(Orgnr("971800534"))
        assertThat(result).isEqualTo(EnhetsregisteretClient.HentUnderenhetFeil.FeilVedDekodingAvJson.left())
    }

    @Test
    fun informasjonOmUnderenhet__skal_feile_hvis_returnert_orgnr_ikke_matcher_med_medsendt_orgnr() {
        stubFor(
            get("/enhetsregisteret/underenheter/777777777").willReturn(
                okJson(
                    """
                    {
                      "organisasjonsnummer": "971800534",
                      "navn": "NAV ARBEID OG YTELSER AVD OSLO",
                      "overordnetEnhet": "999263550",
                      "naeringskode1": {
                        "kode": "84300",
                        "beskrivelse": "Trygdeordninger underlagt offentlig forvaltning"
                      },
                      "antallAnsatte": 40
                    }
                    """.trimIndent()
                )
            )
        )
        enhetsregisteretClient.hentUnderenhet(Orgnr("777777777")).fold(
            {
                assertThat(it).isEqualTo(EnhetsregisteretClient.HentUnderenhetFeil.OrgnrMatcherIkke)
            },
            {
                fail("Skulle feilet")
            }
        )
    }

    companion object {
        private var port = 0

        @BeforeAll
        @JvmStatic
        fun beforeAll(wmRuntimeInfo: WireMockRuntimeInfo) {
            port = wmRuntimeInfo.httpPort
        }
    }
}